### ISSUE TITLE
Allow lines with only whitespace to be treated as "empty"

### ISSUE THREAD

**Issue by chenryhen** [stylelint/stylelint#2435]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug where empty lines with whitespace still trigger an error of "Expected empty line before rule" even if Stylelint is configured to ignore whitespace in empty lines.

> Which rule, if any, is this issue related to?

`no-eol-whitespace` and `rule-empty-line-before`

> What CSS is needed to reproduce this issue?

```css
.foo {
  color: pink;
}
. . .
.bar {
  color: green;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "no-eol-whitespace": [true, {
      "ignore": ["empty-lines"]
    }],
    "rule-empty-line-before": "always"
  }
}
```

> Which version of stylelint are you using?

`7.9.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

plug-in for webpack

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

no

> What did you expect to happen?

No "expected empty line before rules" errors

The reasoning being that if I'm explicitly ignoring whitespace in empty lines, I would expect Stylelint to consider those empty lines as "empty" for all other rules. Otherwise, you need to go in and delete the whitespace anyway.

> What actually happened (e.g. what warnings or errors you are getting)?

See above. Deleting the whitespace in the empty line above the rule removes the ensuing "Expected empty line before rule" error.

**Comment by jeddy3** [comment]
@chenryhen Thanks for the report and for using the template.

I think `no-eol-whitespace`'s `"ignore": ["empty-lines"]` option is a pretty recent addition to stylelint and before that we didn't really have a consistent notion of what "empty" meant. 

Now that we have defined "empty" to be "contains nothing _or only whitespace_", it seems sensible to update all the `*-empty-line-before` rules to consider a line with _only whitespace on_ to be empty.

Additionally, as it currently stands, I think there is a slight overlap between the `no-eol-whitespace` rule and the `*-empty-line-before` rules as the `*-empty-line-before` rules are _indirectly_ checking for trailing whitespace on "empty" lines.

@stylelint/core Any objections to changing the behaviour of the `*-empty-line-before` rules? I believe this change would _loosen_ the restrictiveness of the rules and would not a breaking change. 

> The reasoning being that if I'm explicitly ignoring whitespace in empty lines, I would expect Stylelint to consider those empty lines as "empty" for all other rules.

FYI, all rules in stylelint are standalone and so the options for one rule will not affect the behaviour of another rule. This is an intentional design choice.

**Comment by hudochenkov** [comment]
> Any objections to changing the behaviour of the *-empty-line-before rules? I believe this change would loosen the restrictiveness of the rules and would not a breaking change.

Personally I'm against introducing non-optional looseness. But I don't have any strong arguments against it ¯﻿\﻿_﻿(﻿ツ﻿)﻿_﻿/﻿¯ 

To make this change it's only require to change one file [`hasEmptyLine.js`](https://github.com/stylelint/stylelint/blob/master/lib/utils/hasEmptyLine.js) (and tests).

**Comment by jeddy3** [comment]
> Personally I'm against introducing non-optional looseness.

I see what you're saying but the looseness is covered by `no-eol-whitespace: true`. Also, we're getting a consistent definition of "empty" out of the deal.

> To make this change it's only require to change one file hasEmptyLine.js (and tests).

Good stuff. It looks like extracting big chunks of the empty line and preceding node code into shared utils is going to pay off :)

**Comment by hudochenkov** [comment]
Milestoning this change as `8.0.0`, because it's a breaking change for `never*` options. E. g. for `"rule-empty-line-before": "never"` this was not a warning `b {}\n \na {}` because of space.

**Comment by hudochenkov** [comment]
Thank you, @chenryhen, for bringing this up. New behaviour will be available in `8.0.0`. No ETA, though.

Closed by #2440.
